scripts/perf_report.py

```diff
@@ -277,6 +316,412 @@ def gen_perf_report(heading:list, metrics:dict, base_svr_value:int, base_cli_val
+    the metrics gathered at run-time. Generates a summary performance
+    comparison report.
+
+    The input file is expected to have multiple lines of this format:
+
+Summary: For 32 clients, Baseline - No logging, num_threads=1, num_ops=32000000 (32 Million) ops, Elapsed time=105613818950 (~105.61 Billion) ns, Avg. Elapsed real time=3300 ns/msg, Server throughput=302990 (~302.99 K) ops/sec, Client throughput=10035 (~10.03 K) ops/sec, Avg. ops per thread=32000000 (32 Million) ops/thread
+
+Summary: For 32 clients, Baseline - No logging, num_threads=2, num_ops=32000000 (32 Million) ops, Elapsed time=69448583392 (~69.44 Billion) ns, Avg. Elapsed real time=2170 ns/msg, Server throughput=460772 (~460.77 K) ops/sec, Client throughput=15524 (~15.52 K) ops/sec, Avg. ops per thread=16000000 (16 Million) ops/thread
+
+    The layout of this file is as follows:
+    For a given # of clients: e.g., 'For 32 clients':
+
+     - One summary-line:
+        - for each logging-type
+        - for one server-thread-count
+        - for 1 or more iterations of server execution
+
+     - For one run-parameter: (logging-type, num-server-threads), we could have
+       1 or more lines, for each iteration of the test. (Iteration lines are not
+       tagged.)
+
+     - After n-iterations, we could move to the next num-server-threads count.
+
+     - If the experiment was run with just one thread-count; i.e., env-var
+       L3_PERF_SERVER_NUM_THREADS=4, we will have just one set of these rows
+       for a single thread-count.
+
+    We have to rearrange the metrics rows so that the performance comparison
+    report for each logging-type as compared to baseline performance is
+    generated for different server-thread counts.
+
+    For runs with multiple iterations, this method computes the median metric
+    value and re-casts the metrics to drive-off of that. We massage the computed
+    median metric(s) to invoke existing report-generation code.
+
+    """
+    # pylint: enable=line-too-long
+
+    # Create an in-memory list of lines, which we'll scan multiple times.
+    flines = read_summary_lines(perf_summ_file)
+
+    if len(flines) == 0:
+        return
+
+    (num_clients_msg, num_ops_msg) = extract_nclients_ops_fields(flines[0])
+
+    (svr_metric_name, cli_metric_name, thread_metric_name) = extract_metric_names(flines[0])
+    heading_row = [ 'Run-Type', svr_metric_name, cli_metric_name, thread_metric_name ]
+
+    # Returns 'Baseline' run-type as 0'th row.
+    list_of_nthreads = gen_nthreads_list(flines)
+    # DEBUG: print(list_of_nthreads)
+
+    list_of_runtypes = gen_runtypes_list(flines)
+    # DEBUG: print(list_of_runtypes)
+
+    base_svr_value = 0
+    base_cli_value = 0
+
+    # Run through diff server-thread parameters, and generate report
+    # We make multiple passes extracing metrics for each combination
+    # of #-of-server-threads and run-types.
+    for threads in list_of_nthreads:
```

Comment:
This new approach simplifies the post-processing.

We make multiple passes on the output lines, looking for lines of interest to consolidate and to generate aggregate metrics from.